README.md
# FORK: Use Cryptography instead of M2Crypto
This fork is updated to use the cryptography library for easy install on windows. In addition, m2crypto seems to be unmaintained.

## Installing this fork
You can install this fork like this:
```
pip install git+https://github.com/shivaRamdeen/passbook.git
```

# Passbook

[![Build Status](https://travis-ci.org/devartis/passbook.svg?branch=master)](https://travis-ci.org/devartis/passbook)

Python library to create Apple Wallet (.pkpass) files (Apple Wallet 
has previously been known as Passbook in iOS 6 to iOS 8).

See the [Wallet Topic Page](https://developer.apple.com/wallet/) and the
[Wallet Developer Guide](https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/PassKit_PG/index.html#//apple_ref/doc/uid/TP40012195) for more information about Apple Wallet.

> If you need the server side implementation (API / WebServices) in django you should check http://github.com/devartis/django-passbook.


## Getting Started

1) Get a Pass Type Id

* Visit the iOS Provisioning Portal -> Pass Type IDs -> New Pass Type ID
* Select pass type id -> Configure (Follow steps and download generated pass.cer file)
* Use Keychain tool to export a Certificates.p12 file (need Apple Root Certificate installed)

2) Generate the necessary certificate

```shell
    $ openssl pkcs12 -in "Certificates.p12" -clcerts -nokeys -out certificate.pem   
```
3) Generate the key.pem

```shell
    $ openssl pkcs12 -in "Certificates.p12" -nocerts -out private.key
```

You will be asked for an export password (or export phrase). In this example it will be `123456`, the script will use this as an argument to output the desired `.pkpass`

4) Ensure you have M2Crypto installed

    sudo easy_install M2Crypto

## Typical Usage

    #!/usr/bin/env python

    from passbook.models import Pass, Barcode, StoreCard

    cardInfo = StoreCard()
    cardInfo.addPrimaryField('name', 'John Doe', 'Name')

    organizationName = 'Your organization' 
    passTypeIdentifier = 'pass.com.your.organization' 
    teamIdentifier = 'AGK5BZEN3E'
    
    passfile = Pass(cardInfo, \
        passTypeIdentifier=passTypeIdentifier, \
        organizationName=organizationName, \
        teamIdentifier=teamIdentifier)
    passfile.serialNumber = '1234567' 
    passfile.barcode = Barcode(message = 'Barcode message')    

    # Including the icon and logo is necessary for the passbook to be valid.
    passfile.addFile('icon.png', open('images/icon.png', 'rb'))
    passfile.addFile('logo.png', open('images/logo.png', 'rb'))
    
    # Create and output the Passbook file (.pkpass)
    password = '123456'
    passfile.create('certificate.pem', 'private.key', 'wwdr.pem', password , 'test.pkpass')

## Note: Getting WWDR Certificate

Certificate is available @ http://developer.apple.com/certificationauthority/AppleWWDRCA.cer

It can be exported from KeyChain into a .pem (e.g. wwdr.pem).

## Testing

You can run the tests with `py.test` or optionally with coverage support 
(install `pytest-cov` first): 

    py.test --cov
    
You can also generate a HTML report of the coverage:

    py.test --cov-report html

You can run the tests against multiple versions of Python by running `tox` 
which you need to install first.

## Credits

Developed by [devartis](http://www.devartis.com).

## Contributors

Martin Bächtold


CHANGES.txt
v0.1.0dev, 09/28/2012 -- Initial release.

v0.1.0, 09/28/2012 -- First Released Version.
  Initial implementation of writing of .pkpass files

v0.2.0, 10/01/2012 -- Bugfixes and json properties serialization

v0.2.1, 10/02/2012 -- Bugfix

v0.2.3, 10/31/2012 -- Better handling of Passbook's optional parameters. Improved documentation.

v1.0.0, 12/10/2012 -- Fixed barcode serialization and unicode paths. 

v1.0.1, 08/05/2015 -- Added new field values and validations

v1.0.2, 07/25/2016 -- Add compatibility with iOS 9 (thanks @mbaechtold)

v1.0.3, 12/14/2020 -- Switch crypto library from m2crypto to cryptography


passbook/test/test_passbook.py
# -*- coding: utf-8 -*-
import json

import pytest
from path import Path

from passbook.models import Barcode, BarcodeFormat, Pass, StoreCard

cwd = Path(__file__).parent

wwdr_certificate = cwd / 'certificates' / 'wwdr_certificate.pem'
certificate = cwd / 'certificates' / 'certificate.pem'
key = cwd / 'certificates' / 'private.key'
password_file = cwd / 'certificates' / 'password.txt'


def create_shell_pass(barcodeFormat=BarcodeFormat.CODE128):
    cardInfo = StoreCard()
    cardInfo.addPrimaryField('name', u'Jähn Doe', 'Name')
    stdBarcode = Barcode('test barcode', barcodeFormat, 'alternate text')
    passfile = Pass(cardInfo, organizationName='Org Name', passTypeIdentifier='Pass Type ID', teamIdentifier='Team Identifier')
    passfile.barcode = stdBarcode
    passfile.serialNumber = '1234567'
    passfile.description = 'A Sample Pass'
    return passfile


def test_basic_pass():
    passfile = create_shell_pass()
    assert passfile.formatVersion == 1
    assert passfile.barcode.format == BarcodeFormat.CODE128
    assert len(passfile._files) == 0

    passfile_json = passfile.json_dict()
    assert passfile_json is not None
    assert passfile_json['suppressStripShine'] == False
    assert passfile_json['formatVersion'] == 1
    assert passfile_json['passTypeIdentifier'] == 'Pass Type ID'
    assert passfile_json['serialNumber'] == '1234567'
    assert passfile_json['teamIdentifier'] == 'Team Identifier'
    assert passfile_json['organizationName'] == 'Org Name'
    assert passfile_json['description'] == 'A Sample Pass'


def test_manifest_creation():
    passfile = create_shell_pass()
    manifest_json = passfile._createManifest(passfile._createPassJson())
    manifest = json.loads(manifest_json)
    assert 'pass.json' in manifest


def test_header_fields():
    passfile = create_shell_pass()
    passfile.passInformation.addHeaderField('header', 'VIP Store Card', 'Famous Inc.')
    pass_json = passfile.json_dict()
    assert pass_json['storeCard']['headerFields'][0]['key'] == 'header'
    assert pass_json['storeCard']['headerFields'][0]['value'] == 'VIP Store Card'
    assert pass_json['storeCard']['headerFields'][0]['label'] == 'Famous Inc.'


def test_secondary_fields():
    passfile = create_shell_pass()
    passfile.passInformation.addSecondaryField('secondary', 'VIP Store Card', 'Famous Inc.')
    pass_json = passfile.json_dict()
    assert pass_json['storeCard']['secondaryFields'][0]['key'] == 'secondary'
    assert pass_json['storeCard']['secondaryFields'][0]['value'] == 'VIP Store Card'
    assert pass_json['storeCard']['secondaryFields'][0]['label'] == 'Famous Inc.'


def test_back_fields():
    passfile = create_shell_pass()
    passfile.passInformation.addBackField('back1', 'VIP Store Card', 'Famous Inc.')
    pass_json = passfile.json_dict()
    assert pass_json['storeCard']['backFields'][0]['key'] == 'back1'
    assert pass_json['storeCard']['backFields'][0]['value'] == 'VIP Store Card'
    assert pass_json['storeCard']['backFields'][0]['label'] == 'Famous Inc.'


def test_auxiliary_fields():
    passfile = create_shell_pass()
    passfile.passInformation.addAuxiliaryField('aux1', 'VIP Store Card', 'Famous Inc.')
    pass_json = passfile.json_dict()
    assert pass_json['storeCard']['auxiliaryFields'][0]['key'] == 'aux1'
    assert pass_json['storeCard']['auxiliaryFields'][0]['value'] == 'VIP Store Card'
    assert pass_json['storeCard']['auxiliaryFields'][0]['label'] == 'Famous Inc.'


def test_code128_pass():
    """
    This test is to create a pass with a new code128 format,
    freezes it to json, then reparses it and validates it defaults
    the legacy barcode correctly
    """
    passfile = create_shell_pass()
    jsonData = passfile._createPassJson()
    thawedJson = json.loads(jsonData)
    assert thawedJson['barcode']['format'] == BarcodeFormat.PDF417
    assert thawedJson['barcodes'][0]['format'] == BarcodeFormat.CODE128


def test_pdf_417_pass():
    """
    This test is to create a pass with a barcode that is valid
    in both past and present versions of IOS
    """
    passfile = create_shell_pass(BarcodeFormat.PDF417)
    jsonData = passfile._createPassJson()
    thawedJson = json.loads(jsonData)
    assert thawedJson['barcode']['format'] == BarcodeFormat.PDF417
    assert thawedJson['barcodes'][0]['format'] == BarcodeFormat.PDF417


def test_files():
    passfile = create_shell_pass()
    passfile.addFile('icon.png', open(cwd / 'static/white_square.png', 'rb'))
    assert len(passfile._files) == 1
    assert 'icon.png' in passfile._files

    manifest_json = passfile._createManifest(passfile._createPassJson())
    manifest = json.loads(manifest_json)
    assert '170eed23019542b0a2890a0bf753effea0db181a' == manifest['icon.png']

    passfile.addFile('logo.png', open(cwd / 'static/white_square.png', 'rb'))
    assert len(passfile._files) == 2
    assert 'logo.png' in passfile._files

    manifest_json = passfile._createManifest(passfile._createPassJson())
    manifest = json.loads(manifest_json)
    assert '170eed23019542b0a2890a0bf753effea0db181a' == manifest['logo.png']


# def test_signing():
#     """
#     This test can only run locally if you provide your personal Apple Wallet
#     certificates, private key and password. It would not be wise to add
#     them to git. Store them in the files indicated below, they are ignored
#     by git.
#     """
#     try:
#         with open(password_file) as file_:
#             password = file_.read().strip()
#     except IOError:
#         password = ''

#     passfile = create_shell_pass()
#     manifest_json = passfile._createManifest(passfile._createPassJson())
#     signature = passfile._sign_manifest(
#         manifest_json,
#         certificate,
#         key,
#         wwdr_certificate,
#         password,
#     )

#     smime = passfile._get_smime(
#         certificate,
#         key,
#         wwdr_certificate,
#         password,
#     )

#     store = X509.X509_Store()
#     try:
#         store.load_info(bytes(wwdr_certificate, encoding='utf8'))
#     except TypeError:
#         store.load_info(str(wwdr_certificate))

#     smime.set_x509_store(store)

#     data_bio = BIO.MemoryBuffer(bytes(manifest_json, encoding='utf8'))

#     # PKCS7_NOVERIFY = do not verify the signers certificate of a signed message.
#     assert smime.verify(signature, data_bio, flags=SMIME.PKCS7_NOVERIFY) == bytes(manifest_json, encoding='utf8')

    # tampered_manifest = bytes('{"pass.json": "foobar"}', encoding='utf8')
    # data_bio = BIO.MemoryBuffer(tampered_manifest)
    # # Verification MUST fail!
    # with pytest.raises(SMIME.PKCS7_Error):
    #     smime.verify(signature, data_bio, flags=SMIME.PKCS7_NOVERIFY)


def test_passbook_creation():
    """
    This test can only run locally if you provide your personal Apple Wallet
    certificates, private key and password. It would not be wise to add
    them to git. Store them in the files indicated below, they are ignored
    by git.
    """
    try:
        with open(password_file) as file_:
            password = file_.read().strip()
    except IOError:
        password = ''

    passfile = create_shell_pass()
    passfile.addFile('icon.png', open(cwd / 'static' / 'white_square.png', 'rb'))
    zipfile = passfile.create(certificate, key, wwdr_certificate, password)


passbook/models.py
# -*- coding: utf-8 -*-
import decimal
import hashlib
import json
import zipfile
from io import BytesIO

from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.serialization import pkcs7


class Alignment:
    LEFT = 'PKTextAlignmentLeft'
    CENTER = 'PKTextAlignmentCenter'
    RIGHT = 'PKTextAlignmentRight'
    JUSTIFIED = 'PKTextAlignmentJustified'
    NATURAL = 'PKTextAlignmentNatural'


class BarcodeFormat:
    PDF417 = 'PKBarcodeFormatPDF417'
    QR = 'PKBarcodeFormatQR'
    AZTEC = 'PKBarcodeFormatAztec'
    CODE128 = 'PKBarcodeFormatCode128'


class TransitType:
    AIR = 'PKTransitTypeAir'
    TRAIN = 'PKTransitTypeTrain'
    BUS = 'PKTransitTypeBus'
    BOAT = 'PKTransitTypeBoat'
    GENERIC = 'PKTransitTypeGeneric'


class DateStyle:
    NONE = 'PKDateStyleNone'
    SHORT = 'PKDateStyleShort'
    MEDIUM = 'PKDateStyleMedium'
    LONG = 'PKDateStyleLong'
    FULL = 'PKDateStyleFull'


class NumberStyle:
    DECIMAL = 'PKNumberStyleDecimal'
    PERCENT = 'PKNumberStylePercent'
    SCIENTIFIC = 'PKNumberStyleScientific'
    SPELLOUT = 'PKNumberStyleSpellOut'


class Field(object):

    def __init__(self, key, value, label=''):

        self.key = key  # Required. The key must be unique within the scope
        self.value = value  # Required. Value of the field. For example, 42
        self.label = label  # Optional. Label text for the field.
        self.changeMessage = ''  # Optional. Format string for the alert text that is displayed when the pass is updated
        self.textAlignment = Alignment.LEFT

    def json_dict(self):
        return self.__dict__


class DateField(Field):

    def __init__(self, key, value, label='', dateStyle=DateStyle.SHORT,
                 timeStyle=DateStyle.SHORT, ignoresTimeZone=False):
        super(DateField, self).__init__(key, value, label)
        self.dateStyle = dateStyle  # Style of date to display
        self.timeStyle = timeStyle  # Style of time to display
        self.isRelative = False  # If true, the labels value is displayed as a relative date
        if ignoresTimeZone:
            self.ignoresTimeZone = ignoresTimeZone

    def json_dict(self):
        return self.__dict__


class NumberField(Field):

    def __init__(self, key, value, label=''):
        super(NumberField, self).__init__(key, value, label)
        self.numberStyle = NumberStyle.DECIMAL  # Style of date to display

    def json_dict(self):
        return self.__dict__


class CurrencyField(NumberField):

    def __init__(self, key, value, label='', currencyCode=''):
        super(CurrencyField, self).__init__(key, value, label)
        self.currencyCode = currencyCode  # ISO 4217 currency code

    def json_dict(self):
        return self.__dict__


class Barcode(object):

    def __init__(self, message, format=BarcodeFormat.PDF417, altText='', messageEncoding='iso-8859-1'):
        self.format = format
        self.message = message  # Required. Message or payload to be displayed as a barcode
        self.messageEncoding = messageEncoding  # Required. Text encoding that is used to convert the message
        if altText:
            self.altText = altText  # Optional. Text displayed near the barcode

    def json_dict(self):
        return self.__dict__


class Location(object):

    def __init__(self, latitude, longitude, altitude=0.0):
        # Required. Latitude, in degrees, of the location.
        try:
            self.latitude = float(latitude)
        except (ValueError, TypeError):
            self.latitude = 0.0
        # Required. Longitude, in degrees, of the location.
        try:
            self.longitude = float(longitude)
        except (ValueError, TypeError):
            self.longitude = 0.0
        # Optional. Altitude, in meters, of the location.
        try:
            self.altitude = float(altitude)
        except (ValueError, TypeError):
            self.altitude = 0.0
        # Optional. Notification distance
        self.distance = None
        # Optional. Text displayed on the lock screen when
        # the pass is currently near the location
        self.relevantText = ''

    def json_dict(self):
        return self.__dict__


class IBeacon(object):
    def __init__(self, proximityuuid, major, minor):
        # IBeacon data
        self.proximityUUID = proximityuuid
        self.major = major
        self.minor = minor

        # Optional. Text message where near the ibeacon
        self.relevantText = ''

    def json_dict(self):
        return self.__dict__


class PassInformation(object):

    def __init__(self):
        self.headerFields = []
        self.primaryFields = []
        self.secondaryFields = []
        self.backFields = []
        self.auxiliaryFields = []

    def addHeaderField(self, key, value, label):
        self.headerFields.append(Field(key, value, label))

    def addPrimaryField(self, key, value, label):
        self.primaryFields.append(Field(key, value, label))

    def addSecondaryField(self, key, value, label):
        self.secondaryFields.append(Field(key, value, label))

    def addBackField(self, key, value, label):
        self.backFields.append(Field(key, value, label))

    def addAuxiliaryField(self, key, value, label):
        self.auxiliaryFields.append(Field(key, value, label))

    def json_dict(self):
        d = {}
        if self.headerFields:
            d.update({'headerFields': [f.json_dict() for f in self.headerFields]})
        if self.primaryFields:
            d.update({'primaryFields': [f.json_dict() for f in self.primaryFields]})
        if self.secondaryFields:
            d.update({'secondaryFields': [f.json_dict() for f in self.secondaryFields]})
        if self.backFields:
            d.update({'backFields': [f.json_dict() for f in self.backFields]})
        if self.auxiliaryFields:
            d.update({'auxiliaryFields': [f.json_dict() for f in self.auxiliaryFields]})
        return d


class BoardingPass(PassInformation):

    def __init__(self, transitType=TransitType.AIR):
        super(BoardingPass, self).__init__()
        self.transitType = transitType
        self.jsonname = 'boardingPass'

    def json_dict(self):
        d = super(BoardingPass, self).json_dict()
        d.update({'transitType': self.transitType})
        return d


class Coupon(PassInformation):

    def __init__(self):
        super(Coupon, self).__init__()
        self.jsonname = 'coupon'


class EventTicket(PassInformation):

    def __init__(self):
        super(EventTicket, self).__init__()
        self.jsonname = 'eventTicket'


class Generic(PassInformation):

    def __init__(self):
        super(Generic, self).__init__()
        self.jsonname = 'generic'


class StoreCard(PassInformation):

    def __init__(self):
        super(StoreCard, self).__init__()
        self.jsonname = 'storeCard'


class Pass(object):

    def __init__(self, passInformation, json='', passTypeIdentifier='',
                 organizationName='', teamIdentifier=''):

        self._files = {}  # Holds the files to include in the .pkpass
        self._hashes = {}  # Holds the SHAs of the files array

        # Standard Keys

        # Required. Team identifier of the organization that originated and
        # signed the pass, as issued by Apple.
        self.teamIdentifier = teamIdentifier
        # Required. Pass type identifier, as issued by Apple. The value must
        # correspond with your signing certificate. Used for grouping.
        self.passTypeIdentifier = passTypeIdentifier
        # Required. Display name of the organization that originated and
        # signed the pass.
        self.organizationName = organizationName
        # Required. Serial number that uniquely identifies the pass.
        self.serialNumber = ''
        # Required. Brief description of the pass, used by the iOS
        # accessibility technologies.
        self.description = ''
        # Required. Version of the file format. The value must be 1.
        self.formatVersion = 1

        # Visual Appearance Keys
        self.backgroundColor = None  # Optional. Background color of the pass
        self.foregroundColor = None  # Optional. Foreground color of the pass,
        self.labelColor = None  # Optional. Color of the label text
        self.logoText = None  # Optional. Text displayed next to the logo
        self.barcode = None  # Optional. Information specific to barcodes.  This is deprecated and can only be set to original barcode formats.
        self.barcodes = None #Optional.  All supported barcodes
        # Optional. If true, the strip image is displayed
        self.suppressStripShine = False

        # Web Service Keys

        # Optional. If present, authenticationToken must be supplied
        self.webServiceURL = None
        # The authentication token to use with the web service
        self.authenticationToken = None

        # Relevance Keys

        # Optional. Locations where the pass is relevant.
        # For example, the location of your store.
        self.locations = None
        # Optional. IBeacons data
        self.ibeacons = None
        # Optional. Date and time when the pass becomes relevant
        self.relevantDate = None

        # Optional. A list of iTunes Store item identifiers for
        # the associated apps.
        self.associatedStoreIdentifiers = None
        self.appLaunchURL = None
        # Optional. Additional hidden data in json for the passbook
        self.userInfo = None

        self.expirationDate = None
        self.voided = None

        self.passInformation = passInformation

    # Adds file to the file array
    def addFile(self, name, fd):
        self._files[name] = fd.read()

    # Creates the actual .pkpass file
    def create(self, certificate, key, wwdr_certificate, password, zip_file=None):
        pass_json = self._createPassJson()
        manifest = self._createManifest(pass_json)
        signature = self._createSignatureCrypto(manifest, certificate, key, wwdr_certificate, password)
        # signature = self._createSignature(manifest, certificate, key, wwdr_certificate, password)
        if not zip_file:
            zip_file = BytesIO()
        self._createZip(pass_json, manifest, signature, zip_file=zip_file)
        return zip_file

    def _createPassJson(self):
        return json.dumps(self, default=PassHandler)

    def _createManifest(self, pass_json):
        """
        Creates the hashes for all the files included in the pass file.
        """
        self._hashes['pass.json'] = hashlib.sha1(pass_json.encode('utf-8')).hexdigest()
        for filename, filedata in self._files.items():
            self._hashes[filename] = hashlib.sha1(filedata).hexdigest()
        return json.dumps(self._hashes)

    # def _get_smime(self, certificate, key, wwdr_certificate, password):
    #     """
    #     :return: M2Crypto.SMIME.SMIME
    #     """
    #     def passwordCallback(*args, **kwds):
    #         return bytes(password, encoding='ascii')

    #     smime = SMIME.SMIME()

    #     wwdrcert = X509.load_cert(wwdr_certificate)
    #     stack = X509_Stack()
    #     stack.push(wwdrcert)
    #     smime.set_x509_stack(stack)

    #     smime.load_key(key, certfile=certificate, callback=passwordCallback)
    #     return smime

    # def _sign_manifest(self, manifest, certificate, key, wwdr_certificate, password):
    #     """
    #     :return: M2Crypto.SMIME.PKCS7
    #     """
    #     smime = self._get_smime(certificate, key, wwdr_certificate, password)
    #     pkcs7 = smime.sign(
    #         SMIME.BIO.MemoryBuffer(bytes(manifest, encoding='utf8')),
    #         flags=SMIME.PKCS7_DETACHED | SMIME.PKCS7_BINARY
    #     )
    #     return pkcs7

    # def _createSignature(self, manifest, certificate, key,
    #                      wwdr_certificate, password):
    #     """
    #     Creates a signature (DER encoded) of the manifest. The manifest is the file
    #     containing a list of files included in the pass file (and their hashes).
    #     """
    #     pk7 = self._sign_manifest(manifest, certificate, key, wwdr_certificate, password)
    #     der = SMIME.BIO.MemoryBuffer()
    #     pk7.write_der(der)
    #     return der.read()
    
    def _readFileBytes(self, path):
        """
        Utility function to read files as byte data
        :param path: file path
        :returns bytes
        """
        file = open(path)
        return file.read().encode('UTF-8')

    def _createSignatureCrypto(self, manifest, certificate, key,
                         wwdr_certificate, password):
        """
        Creates a signature (DER encoded) of the manifest.
        Rewritten to use cryptography library instead of M2Crypto 
        The manifest is the file
        containing a list of files included in the pass file (and their hashes).
        """
        cert = x509.load_pem_x509_certificate(self._readFileBytes(certificate))
        priv_key = serialization.load_pem_private_key(self._readFileBytes(key), password=password.encode('UTF-8'))
        wwdr_cert = x509.load_pem_x509_certificate(self._readFileBytes(wwdr_certificate))
        
        options = [pkcs7.PKCS7Options.DetachedSignature]
        return pkcs7.PKCS7SignatureBuilder()\
                .set_data(manifest.encode('UTF-8'))\
                .add_signer(cert, priv_key, hashes.SHA1())\
                .add_certificate(wwdr_cert)\
                .sign(serialization.Encoding.DER, options)

    # Creates .pkpass (zip archive)
    def _createZip(self, pass_json, manifest, signature, zip_file=None):
        zf = zipfile.ZipFile(zip_file or 'pass.pkpass', 'w')
        zf.writestr('signature', signature)
        zf.writestr('manifest.json', manifest)
        zf.writestr('pass.json', pass_json)
        for filename, filedata in self._files.items():
            zf.writestr(filename, filedata)
        zf.close()

    def json_dict(self):
        d = {
            'description': self.description,
            'formatVersion': self.formatVersion,
            'organizationName': self.organizationName,
            'passTypeIdentifier': self.passTypeIdentifier,
            'serialNumber': self.serialNumber,
            'teamIdentifier': self.teamIdentifier,
            'suppressStripShine': self.suppressStripShine,
            self.passInformation.jsonname: self.passInformation.json_dict()
        }
        #barcodes have 2 fields, 'barcode' is legacy so limit it to the legacy formats, 'barcodes' supports all
        if self.barcode:
            original_formats = [BarcodeFormat.PDF417, BarcodeFormat.QR, BarcodeFormat.AZTEC]
            legacyBarcode = self.barcode
            newBarcodes = [self.barcode.json_dict()]
            if self.barcode.format not in original_formats:
                legacyBarcode = Barcode(self.barcode.message, BarcodeFormat.PDF417, self.barcode.altText)
            d.update({'barcodes': newBarcodes})
            d.update({'barcode': legacyBarcode})

        if self.relevantDate:
            d.update({'relevantDate': self.relevantDate})
        if self.backgroundColor:
            d.update({'backgroundColor': self.backgroundColor})
        if self.foregroundColor:
            d.update({'foregroundColor': self.foregroundColor})
        if self.labelColor:
            d.update({'labelColor': self.labelColor})
        if self.logoText:
            d.update({'logoText': self.logoText})
        if self.locations:
            d.update({'locations': self.locations})
        if self.ibeacons:
            d.update({'beacons': self.ibeacons})
        if self.userInfo:
            d.update({'userInfo': self.userInfo})
        if self.associatedStoreIdentifiers:
            d.update(
                {'associatedStoreIdentifiers': self.associatedStoreIdentifiers}
            )
        if self.appLaunchURL:
            d.update({'appLaunchURL': self.appLaunchURL})
        if self.expirationDate:
            d.update({'expirationDate': self.expirationDate})
        if self.voided:
            d.update({'voided': True})
        if self.webServiceURL:
            d.update({'webServiceURL': self.webServiceURL,
                      'authenticationToken': self.authenticationToken})
        return d


def PassHandler(obj):
    if hasattr(obj, 'json_dict'):
        return obj.json_dict()
    else:
        # For Decimal latitude and logitude etc.
        if isinstance(obj, decimal.Decimal):
            return str(obj)
        else:
            return obj


passbook/passbook.py
import json
import hashlib
import zipfile
import base64
import os.path
from M2Crypto import SMIME

class PKPass :
    
    certPath = '' # Holds the path to the certificate    
    certPass = '' # Holds the password to the certificate    
    files = [] # Holds the files to include in the .pkpass    
    JSON = '' # Holds the json    
    SHAs = {} # Holds the SHAs of the files array

    def __init__(self, certPath = '', certPass = '', JSON = ''):

        self.cert = certPath
        self.certPass = certPass
        self.JSON = JSON
    
        
    # Adds file to the file array
    def addFile(self, path):

        if os.path.exists(path):
            self.files.append(path)

    
    # Creates the actual .pkpass file
    def create(self):

        manifest = self.createManifest()
        self.createSignature(manifest)
        self.createZip(manifest)
        
    
    # creates the hashes for the files and adds them into a json string.
    def createManifest(self):

        # Creates SHA hashes for all files in package
        self.SHAs['pass.json'] = hashlib.sha1(self.JSON).hexdigest()
        for filename in self.files :
            self.SHAs[os.path.basename(filename)] = hashlib.sha1(open(filename).read()).hexdigest()
        
        manifest = json.dumps(self.SHAs)
        
        return manifest
    
    
    # Creates a signature and saves it
    def createSignature(self, manifest):

        open('manifest.json', 'w').write(manifest)        

        def passwordCallback(*args, **kwds):
            return self.certPass

        smime = SMIME.SMIME()
        smime.load_key('key.pem', 'certificate.pem', callback=passwordCallback)        
        pk7 = smime.sign(SMIME.BIO.MemoryBuffer(manifest), flags=SMIME.PKCS7_BINARY)                
        pem = SMIME.BIO.MemoryBuffer()
        pk7.write(pem)
        # convert pem to der
        der = ''.join(l.strip() for l in pem.read().split('-----')[2].splitlines()).decode('base64')        

        open('signature', 'w').write(der)
        
    
    # Creates .pkpass (zip archive)
    def createZip(self, manifest):

        # Package file in Zip (as .pkpass)
        with zipfile.ZipFile('pass.pkpass', 'w') as zf:
            zf.write('signature', 'signature')
            zf.writestr('manifest.json', manifest)
            zf.writestr('pass.json', self.JSON)
            for filename in self.files:
                zf.write(filename, os.path.basename(filename))
    
    
def main():

    pkfile = PKPass()
    pkfile.certPath = 'Certificates.p12' # Set the path to your Pass Certificate (.p12 file)
    pkfile.certPass = '123456' # Set password for certificate
    pkfile.JSON = '{ \
        "passTypeIdentifier": "pass.com.devartis.test", \
        "formatVersion": 1, \
        "organizationName": "devartis", \
        "serialNumber": "123456", \
        "teamIdentifier": "AGK5BZEN3E", \
        "backgroundColor": "rgb(107,156,196)", \
        "logoText": "devartis", \
        "description": "Demo pass", \
    	"boardingPass": { \
            "primaryFields": [ \
                { \
                    "key" : "origin", \
                	"label" : "Buenos Aires", \
                	"value" : "EZE" \
                }, \
                { \
                	"key" : "destination", \
                	"label" : "Mar del plata", \
                	"value" : "MDQ" \
                } \
            ], \
            "secondaryFields": [ \
                { \
                    "key": "gate", \
                    "label": "Gate", \
                    "value": "F12" \
                }, \
                { \
                    "key": "date", \
                    "label": "Departure date", \
                    "value": "08/11/2012 10:22" \
                } \
             ], \
            "backFields": [ \
                { \
                    "key": "passenger-name", \
                    "label": "Passenger", \
                    "value": "Fernando Aramendi" \
                } \
            ], \
            "transitType" : "PKTransitTypeAir" \
        }, \
        "barcode": { \
            "format": "PKBarcodeFormatQR", \
            "message": "Flight-GateF12-ID6643679AH7B", \
            "messageEncoding": "iso-8859-1" \
        } \
    }'

    # add files to the PKPass package
    pkfile.addFile('images/icon.png')
    pkfile.addFile('images/icon@2x.png')
    pkfile.addFile('images/logo.png')
    pkfile.create() # Create and output the PKPass

if __name__ == "__main__":
    main()